imap/caldav_alarm.c

```diff
@@ -2163,6 +2178,400 @@ EXPORTED int caldav_alarm_process(time_t runtime, time_t *intervalp, int dryrun)
+
+    /* Skip declined events */
+    get_schedule_addresses(intname, userid, &sched_addrs);
+    for (prop = icalcomponent_get_first_invitee(comp);
+         prop;
+         prop = icalcomponent_get_next_invitee(comp))
+    {
+        const char *attendee = icalproperty_get_invitee(prop);
+
+        if (!attendee) continue;
+        if (!strncasecmp(attendee, "mailto:", 7)) attendee += 7;
+
+        if (strarray_find_case(&sched_addrs, attendee, 0) >= 0) {
+            const char *partstat =
+                icalproperty_get_parameter_as_string(prop, "PARTSTAT");
+
+            if (!strcasecmpsafe(partstat, "DECLINED")) {
+                strarray_fini(&sched_addrs);
+                return 1;
+            }
+        }
+    }
+    strarray_fini(&sched_addrs);
+
+    for (alarm = icalcomponent_get_first_component(comp, ICAL_VALARM_COMPONENT);
+         alarm;
+         alarm = icalcomponent_get_next_component(comp, ICAL_VALARM_COMPONENT))
+    {
+        prop = icalcomponent_get_first_property(alarm, ICAL_ACTION_PROPERTY);
+        if (!prop) {
+            /* no action, invalid alarm, skip */
+            continue;
+        }
+
+        val = icalproperty_get_value(prop);
+        enum icalproperty_action action = icalvalue_get_action(val);
+        if (!(action == ICAL_ACTION_DISPLAY || action == ICAL_ACTION_EMAIL)) {
+            /* we only want DISPLAY and EMAIL, skip */
+            continue;
+        }
+
+        prop = icalcomponent_get_first_property(alarm, ICAL_TRIGGER_PROPERTY);
+        if (!prop) {
+            /* no trigger, invalid alarm, skip */
+            continue;
+        }
+
+        val = icalproperty_get_value(prop);
+
+        struct icaltriggertype trigger = icalvalue_get_trigger(val);
+        /* XXX validate trigger */
+
+        icaltimetype alarmtime = icaltime_null_time();
+        /* XXX do we want alarmtime? */ (void) alarmtime;
+        unsigned is_duration = (icalvalue_isa(val) == ICAL_DURATION_VALUE);
+        if (is_duration) {
+            icalparameter *param =
+                icalproperty_get_first_parameter(prop, ICAL_RELATED_PARAMETER);
+            icaltimetype base = start;
+            if (param && icalparameter_get_related(param) == ICAL_RELATED_END) {
+                base = end;
+            }
+            base.is_date = 0; /* need an actual time for triggers */
+            alarmtime = icaltime_add(base, trigger.duration);
+        }
+        else {
+            /* absolute */
+            alarmtime = trigger.time;
+        }
+        alarmtime.is_date = 0;
+
+//        send_alarm(data, comp, alarm, start, end, recurid, is_standalone, alarmtime);
+//    FILL_STRING_PARAM(event, EVENT_CALENDAR_ALARM_TIME,
+//                      xstrdup(icaltime_as_ical_string(alarmtime)));
+
+        /* XXX what other fields are useful here? alarmtime? */
+        lrock->proc(lrock->mbname, lrock->data->imap_uid,
+                    lrock->data->nextcheck, lrock->data->num_rcpts,
```

Comment:
This should be `check` instead of `lrock->data->nextcheck`. I guess that the `num_retries`, `last_run` and `last_err` fields should be zero for future instances, e.g. if `check != lrock->data->nextcheck`, e.g. there can't be a last error for these, yet.